File: 2019/Elixir/2/main.exs
require File
require IO

defmodule Main do

    # Part 1
    def runIntsProgram(verb \\ 12, noun \\ 2) do
        case File.read "input.txt" do
            {:ok, contents} -> 
                contents
                    |> String.trim 
                    |> String.split(",", trim: true)
                    |> Enum.map(&String.to_integer/1)
                    |> fn l ->
                            restoredMemory = List.replace_at(l, 1, verb) |> List.replace_at(2, noun)
                            runInstructions(restoredMemory, restoredMemory) end.()
        
            {:error, reason} ->
                IO.puts("Cannot open input.txt: #{reason}")
        end
    end

    # Part 2
    def findNounAndVerb() do
        case File.read "input.txt" do
            {:ok, contents} -> 
                contents
                    |> String.trim 
                    |> String.split(",", trim: true)
                    |> Enum.map(&String.to_integer/1)
                    |> fn l ->
                            for verb <- 0..99,
                                noun <- 0..99 do
                                restoredMemory = List.replace_at(l, 1, noun) |> List.replace_at(2, verb)
                                result = List.first(runInstructions(restoredMemory, restoredMemory))
                                if result === 19690720 do
                                    IO.puts("Found it! #{100 * noun + verb}")
                                end
                                result
                            end 
                       end.()
        
            {:error, reason} ->
                IO.puts("Cannot open input.txt: #{reason}")
        end
    end 

    def test() do
        program = [1,9,10,3,2,3,11,0,99,30,40,50]
        memory = program 
        IO.inspect(runInstructions(program, memory))
    end
    
    def runInstructions([], memory), do: memory 
    def runInstructions([opcode | tail], memory) do
        case opcode do
            1 -> 
                [operand1Index, operand2Index, storeIndex] = Enum.take(tail, 3)
                operand1 = Enum.at(memory, operand1Index)
                operand2 = Enum.at(memory, operand2Index)
                alteredMemory = List.replace_at(memory, storeIndex, operand1 + operand2)
                program = Enum.drop(tail, 3) 
                runInstructions(program, alteredMemory)
            2 ->
                [operand1Index, operand2Index, storeIndex] = Enum.take(tail, 3)
                operand1 = Enum.at(memory, operand1Index)
                operand2 = Enum.at(memory, operand2Index)
                alteredMemory = List.replace_at(memory, storeIndex , operand1 * operand2)
                program = Enum.drop(tail, 3) 
                runInstructions(program, alteredMemory)
            99 ->
                memory
        end
    end    
     
end

File: 2019/Haskell/2/main.hs
#!/usr/bin/env stack
-- stack --system-ghc --resolver lts-14.16 script

{-# LANGUAGE OverloadedStrings #-} 

import Control.Applicative ((<**>))

import Data.List
import qualified Data.Text    as Text
import qualified Data.Text.IO as Text

main :: IO ()
main = Text.readFile "input.txt"
    <**> return (((map $ read . Text.unpack) . (concatMap $ Text.splitOn ",") . Text.lines) :: Text.Text -> [Integer])         
    >>= \program -> do
            let restoredMemory = setAt 2 2 $ setAt 1 12 program
            putStrLn $ "Restored Result: " ++ show (head $ runProgram restoredMemory restoredMemory)
        
            let (noun, verb) = head $ dropWhile (\(noun, verb) ->
                                                    let restoredMemory = setAt 1 noun $ setAt 2 verb program in
                                                    head (runProgram restoredMemory restoredMemory) /= 19690720
                                                ) ((,) <$> [0..99] <*> [0..99])
            
            putStrLn $ "100 * " ++ show noun ++ " + " ++ show verb ++ " is: " ++ show (100 * noun + verb)

        

test :: [Integer]
test = let program = [1,9,10,3,2,3,11,0,99,30,40,50]
           memory = program
           in runProgram program memory

runProgram :: [Integer] -> [Integer] -> [Integer]
runProgram [] memory = memory
runProgram program@(opcode:tail) memory =
    case opcode of
        1 ->
            let [parameter1Index, parameter2Index, storeIndex] = take 3 tail
                parameter1 = memory `genericIndex` parameter1Index
                parameter2 = memory `genericIndex` parameter2Index
                alteredMemory = setAt storeIndex (parameter1 + parameter2) memory 
                remainingProgram = drop 3 tail
                in runProgram remainingProgram alteredMemory 
        2 -> 
            let [parameter1Index, parameter2Index, storeIndex] = take 3 tail
                parameter1 = memory `genericIndex` parameter1Index
                parameter2 = memory `genericIndex` parameter2Index
                alteredMemory = setAt storeIndex (parameter1 * parameter2) memory 
                remainingProgram = drop 3 tail
                in runProgram remainingProgram alteredMemory 
        99 ->
            memory
        n -> error $ "Unrecognized opcode " ++ show n


setAt :: Integer -> a -> [a] -> [a]
setAt index value list = genericTake index list  ++ value:genericDrop (index + 1) list
